near/src/runtime.rs

```diff
@@ -854,66 +853,42 @@ impl RuntimeAdapter for NightshadeRuntime {
-        if part_id > 0 {
-            /* TODO MOO */
-            return Ok((StatePart { shard_id, part_id, data: vec![] }, vec![]));
-        }
-        // TODO(1052): make sure state_root is present in the trie.
-        // create snapshot.
-        let mut result = vec![];
-        let mut cursor = Cursor::new(&mut result);
-        for item in self.trie.iter(&state_root.hash)? {
-            let (key, value) = item?;
-            cursor.write_u32::<LittleEndian>(key.len() as u32)?;
-            cursor.write_all(&key)?;
-            cursor.write_u32::<LittleEndian>(value.len() as u32)?;
-            cursor.write_all(value.as_ref())?;
-        }
-        // TODO(1048): Save on disk an snapshot, split into chunks and compressed. Send chunks instead of single blob.
-        debug!(target: "runtime", "Read state part #{} for shard #{} @ {}, size = {}", part_id, shard_id, state_root.hash, result.len());
-        // TODO add proof in Nightshade Runtime
-        Ok((StatePart { shard_id, part_id, data: result }, vec![]))
+        part_id: u64,
+        num_parts: u64,
+    ) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
+        let data = self
+            .trie
+            .get_trie_nodes_for_part(part_id, num_parts, state_root)?
+            .try_to_vec()
+            .expect("Failed to serialize state part");
+        Ok(data)
     }
 
-    fn accept_state_part(
+    fn validate_state_part(
         &self,
         state_root: &StateRoot,
-        part: &StatePart,
-        _proof: &Vec<u8>,
+        part_id: u64,
+        num_parts: u64,
+        data: &Vec<u8>,
     ) -> Result<(), Box<dyn std::error::Error>> {
-        if part.part_id > 0 {
-            /* TODO MOO */
-            return Ok(());
-        }
-        debug!(target: "runtime", "Writing state part #{} for shard #{} @ {}, size = {}", part.part_id, part.shard_id, state_root.hash, part.data.len());
-        // TODO prove that the part is valid
-        let mut state_update = TrieUpdate::new(self.trie.clone(), CryptoHash::default());
-        let state_part_len = part.data.len();
-        let mut cursor = Cursor::new(part.data.clone());
-        while cursor.position() < state_part_len as u64 {
-            let key_len = cursor.read_u32::<LittleEndian>()? as usize;
-            let mut key = vec![0; key_len];
-            cursor.read_exact(&mut key)?;
-            let value_len = cursor.read_u32::<LittleEndian>()? as usize;
-            let mut value = vec![0; value_len];
-            cursor.read_exact(&mut value)?;
-            state_update.set(key, DBValue::from_slice(&value));
-        }
-        let (store_update, root) = state_update.finalize()?.into(self.trie.clone())?;
-        if root != state_root.hash {
-            return Err("Invalid state root".into());
-        }
-        store_update.commit()?;
-        Ok(())
+        let trie_nodes: PartialStateStruct =
+            BorshDeserialize::try_from_slice(data).expect("Failed to deserialize state part");
```

Comment:
We cannot panic on invalid data here, it could come from a malicious peer

Reply:
++